src/Cake.Core/Tooling/ToolResolutionStrategy.cs

```diff
@@ -0,0 +1,148 @@
+        private readonly ICakeEnvironment _environment;
+        private readonly IGlobber _globber;
+        private readonly ICakeConfiguration _configuration;
+        private readonly object _lock;
+        private List<DirectoryPath> _path;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ToolResolutionStrategy"/> class.
+        /// </summary>
+        /// <param name="fileSystem">The file system.</param>
+        /// <param name="environment">The environment.</param>
+        /// <param name="globber">The globber.</param>
+        /// <param name="configuration">The configuration.</param>
+        public ToolResolutionStrategy(
+            IFileSystem fileSystem,
+            ICakeEnvironment environment,
+            IGlobber globber,
+            ICakeConfiguration configuration)
+        {
+            if (fileSystem == null)
+            {
+                throw new ArgumentNullException("fileSystem");
+            }
+            if (environment == null)
+            {
+                throw new ArgumentNullException("environment");
+            }
+            if (globber == null)
+            {
+                throw new ArgumentNullException("globber");
+            }
+            _fileSystem = fileSystem;
+            _environment = environment;
+            _globber = globber;
+            _configuration = configuration;
+            _lock = new object();
+        }
+
+        /// <summary>
+        /// Resolves the specified tool using the specified tool repository.
+        /// </summary>
+        /// <param name="repository">The tool repository.</param>
+        /// <param name="tool">The tool.</param>
+        /// <returns>
+        /// The path to the tool; otherwise <c>null</c>.
+        /// </returns>
+        public FilePath Resolve(IToolRepository repository, string tool)
+        {
+            if (repository == null)
+            {
+                throw new ArgumentNullException("repository");
+            }
+            if (tool == null)
+            {
+                throw new ArgumentNullException("tool");
+            }
+            if (string.IsNullOrWhiteSpace(tool))
+            {
+                throw new ArgumentException("Tool name cannot be empty.", "tool");
+            }
+
+            // Does this file already have registrations?
+            var resolve = LookInRegistrations(repository, tool);
+            if (resolve == null)
+            {
+                // Look in ./tools/
+                resolve = LookInToolsDirectory(tool);
+                if (resolve == null)
+                {
+                    // Look in the path environment variable.
+                    resolve = LookInPath(tool);
+                }
+            }
+            return resolve;
+        }
+
+        private static FilePath LookInRegistrations(IToolRepository repository, string tool)
+        {
+            return repository.Resolve(tool).LastOrDefault();
```

Comment:
@patriksvensson I think we need to put this caveat into the documentation.  i.e. if the tool is found in multiple places, the last registration wins.